dbms/src/Storages/Transaction/PDTiKVClient.h

```diff
@@ -31,17 +32,57 @@
 
 #include <atomic>
 
+using TimePoint = std::atomic<std::chrono::time_point<std::chrono::steady_clock>>;
+
+struct KeyspaceGCInfo
+{
+    DB::Timestamp ks_gc_sp;
+    TimePoint ks_gc_sp_update_time;
+
+    KeyspaceGCInfo()
+    {
+        ks_gc_sp_update_time = std::chrono::steady_clock::now();
+    }
+
+    KeyspaceGCInfo(const KeyspaceGCInfo & other)
+    {
+        ks_gc_sp = other.ks_gc_sp;
+        ks_gc_sp_update_time = std::chrono::steady_clock::now();
+    }
+
+    KeyspaceGCInfo & operator=(const KeyspaceGCInfo & other)
+    {
+        if (this != &other)
+        {
+            ks_gc_sp = other.ks_gc_sp;
+            ks_gc_sp_update_time = std::chrono::steady_clock::now();
+        }
+        return *this;
+    }
+};
+
 namespace DB
 {
 struct PDClientHelper
 {
     static constexpr int get_safepoint_maxtime = 120000; // 120s. waiting pd recover.
 
+    static bool enable_safepoint_v2;
+
     static Timestamp getGCSafePointWithRetry(
         const pingcap::pd::ClientPtr & pd_client,
+        KeyspaceID keyspace_id,
         bool ignore_cache = true,
         Int64 safe_point_update_interval_seconds = 30)
     {
+        // If keyspace id is `NullspaceID` it need to use safe point v1.
+        if (enable_safepoint_v2 && keyspace_id != NullspaceID)
+        {
+            auto gc_safe_point = getGCSafePointV2WithRetry(pd_client, keyspace_id, ignore_cache, safe_point_update_interval_seconds);
+            LOG_DEBUG(Logger::get(), "use safe point v2, keyspace_id={},gc_safe_point={}", keyspace_id, gc_safe_point);
```

Comment:
```suggestion
            LOG_DEBUG(Logger::get(), "use safe point v2, keyspace={} gc_safe_point={}", keyspace_id, gc_safe_point);
```

Other codes use 'keyspace=xxx' in logging, so this is better for searching logs
https://github.com/pingcap/tiflash/blob/4ecf3cd7649f0597c540cdbfe4ea609f26d7d169/dbms/src/Storages/GCManager.cpp#L69